File: internal/utils/qrcode.go
package utils

import (
	"encoding/base64"
	"fmt"

	"github.com/google/uuid"
	"github.com/skip2/go-qrcode"
)

func GenerateQRCode(userID string) (string, string, error) {
	// Generate unique QR code data
	qrData := fmt.Sprintf("CHATAPP:%s:%s", userID, uuid.New().String())
	
	// Generate QR code image
	qr, err := qrcode.New(qrData, qrcode.Medium)
	if err != nil {
		return "", "", err
	}

	// Convert to PNG bytes
	png, err := qr.PNG(256)
	if err != nil {
		return "", "", err
	}

	// Convert to base64 string for storage
	qrBase64 := base64.StdEncoding.EncodeToString(png)
	
	return qrData, qrBase64, nil
}

func ParseQRCode(qrData string) (string, error) {
	// Parse QR code format: CHATAPP:user_id:uuid
	var userID string
	_, err := fmt.Sscanf(qrData, "CHATAPP:%s", &userID)
	if err != nil {
		return "", err
	}
	return userID, nil
}

File: src/utils/qrcode.js
'use strict';

const QRCode = require('qrcode');
const { v4: uuidv4 } = require('uuid');

async function generateQRCode(userId) {
  const qrData = `CHATAPP:${userId}:${uuidv4()}`;

  const buffer = await QRCode.toBuffer(qrData, {
    errorCorrectionLevel: 'M',
    width: 256,
    type: 'png',
  });

  const qrBase64 = buffer.toString('base64');
  return { qrData, qrBase64 };
}

function parseQRCode(qrData) {
  // Format: CHATAPP:userId:uuid
  if (!qrData || !qrData.startsWith('CHATAPP:')) {
    return { userId: null, error: 'Invalid QR code format' };
  }
  const parts = qrData.split(':');
  if (parts.length < 2) {
    return { userId: null, error: 'Invalid QR code format' };
  }
  return { userId: parts[1], error: null };
}

module.exports = { generateQRCode, parseQRCode };
